### PR TITLE
Product purchase provisioning

### PR DESCRIPTION
**Pull request by philip-dds**
https://www.pivotaltracker.com/n/projects/2160940/stories/170267273

state machine extended to support two new stages 
PRODUCT_PURCHASE and PRODUCT_PURCHASE_VERIFICATION

### CODE REVIEW

File: atst/domain/csp/cloud/azure_cloud_provider.py [hunk 1]
```diff
@@ -493,6 +497,64 @@ def create_billing_instruction(self, payload: BillingInstructionCSPPayload):
...
+            return self._error(result.json())
+
+
+    def create_product_purchase_verification(
+        self, payload: ProductPurchaseVerificationCSPPayload
+    ):
+        sp_token = self._get_sp_token(payload.creds)
+        if sp_token is None:
+            raise AuthenticationException(
+                "Could not resolve token for task order billing validation"
```
Comment (tomdds) [inline atst/domain/csp/cloud/azure_cloud_provider.py]: Exception message should get updated to reflect this operation (AAD Premium Purchase)

File: atst/domain/csp/cloud/azure_cloud_provider.py [hunk 2]
```diff
@@ -493,6 +497,64 @@ def create_billing_instruction(self, payload: BillingInstructionCSPPayload):
         else:
             return self._error(result.json())
 
+    def create_product_purchase(
```
Comment (tomdds) [inline atst/domain/csp/cloud/azure_cloud_provider.py]: Can you add a unit test for this in test_azure_csp.py?

File: atst/domain/csp/cloud/azure_cloud_provider.py [hunk 3]
```diff
@@ -493,6 +497,64 @@ def create_billing_instruction(self, payload: BillingInstructionCSPPayload):
...
+        self, payload: ProductPurchaseCSPPayload
+    ):
+        sp_token = self._get_sp_token(payload.creds)
+        if sp_token is None:
+            raise AuthenticationException(
+                "Could not resolve token for aad premium product purchase"
+            )
+
+        create_product_purchase_body = payload.dict(by_alias=True)
```
Comment (tomdds) [inline atst/domain/csp/cloud/azure_cloud_provider.py]: The body for this call should look like this:
```json
{
    "type": "AADPremium",
    "sku": "AADP1",
    "productProperties": {
        "beneficiaryTenantId": "<benefit_tid>"
    },
    "quantity": 4
}
```

Currently, calling ProductPurchaseCSPPayload.dict(by_alias=True) will produce this, which wouldn't work for the call:
```json
{
    "billingAccountName": "str",
    "billingProfileName": "str",
    "parameters": [
        {
            "type": "str",
            "sku": "str",
            "quantity": "int",
            "productProperties": {}
        }
    ]
}
```

Can you construct the body in this method by selecting from the payload? You can also include the type and sku consts in this function directly: `AADPremium` `AADP1`

File: atst/domain/csp/cloud/azure_cloud_provider.py [hunk 4]
```diff
@@ -517,6 +522,70 @@ def create_billing_instruction(self, payload: BillingInstructionCSPPayload):
...
+            )
+
+        auth_header = {
+            "Authorization": f"Bearer {sp_token}",
+        }
+
+        result = self.sdk.requests.get(
+            payload.product_purchase_verify_url, headers=auth_header
+        )
+        premium_purchase_date = result.json()["product"]["properties"]["purchaseDate"]
```
Comment (dandds) [inline atst/domain/csp/cloud/azure_cloud_provider.py]: Do we get this value in the response regardless of whether it's a 202 or 200?

File: atst/domain/csp/cloud/models.py [hunk 5]
```diff
@@ -406,3 +406,27 @@ def enforce_root_creds(cls, values):
...
+
+class ProductPurchaseCSPResult(AliasModel):
+    product_purchase_verify_url: str
+    product_purchase_retry_after: int
+
+    class Config:
+        fields = {
+            "product_purchase_verify_url": "Location",
+            "product_purchase_retry_after": "Retry-After",
+        }
```
Comment (dandds) [inline atst/domain/csp/cloud/models.py]: I had to look up what this was doing in the pydantic docs, but this is a neat feature.

Comment (philip-dds) [reply atst/domain/csp/cloud/models.py]: yes, this json schema generation is a nice feature.
Tom was talking about composing a final full schema our of these data classes.

File: atst/domain/csp/cloud/azure_cloud_provider.py (hunk 4 above)
Comment (philip-dds) [reply atst/domain/csp/cloud/azure_cloud_provider.py]: actually I am not sure about 200. The swagger docs only mention 202.
I will ask Tom what he thinks.

Comment (tomdds) [reply atst/domain/csp/cloud/azure_cloud_provider.py]: Whoops, fixed!

**Pull request merged**